akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,170 @@
+        }
+
+        var upstreamFailure = OptionVal.none[Throwable]
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          upstreamFailure = OptionVal.Some(ex)
+        }
+
+        override def onUpstreamFinish(): Unit = {
+          //will later be propagated to the materialized flow
+        }
+
+        var hasBeenPulled = false
+        override def onPull(): Unit = {
+          hasBeenPulled = true
+        }
+
+        var downstreamCause = OptionVal.none[Throwable]
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          downstreamCause = OptionVal.Some(cause)
+        }
+
+        def onFuture(futureRes: Try[Flow[In, Out, M]]) = futureRes match {
+          case Failure(exception) =>
+            setKeepGoing(false)
+            innerMatValue.failure(new NeverMaterializedException(exception))
+            failStage(exception)
+          case Success(flow) =>
+            //materialize flow, connect to inlets, feed with potential events and set handlers
+            connect(flow)
+            setKeepGoing(false)
+        }
+      }
+      def connect(flow: Flow[In, Out, M]): Unit = {
+        val subSource = new SubSourceOutlet[In](s"${FutureFlow.this}.subIn")
+        val subSink = new SubSinkInlet[Out](s"${FutureFlow.this}.subOut")
+
+        subSource.setHandler {
+          new OutHandler {
+            override def onPull(): Unit = if (!isClosed(in)) tryPull(in)
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if (!isClosed(in)) {
+                cancel(in, cause)
+              }
+            }
+          }
+        }
+        subSink.setHandler {
+          new InHandler {
+            override def onPush(): Unit = {
+              push(out, subSink.grab())
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              complete(out)
+            }
+
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              fail(out, ex)
+            }
+          }
+        }
+        Try {
+          Source.fromGraph(subSource.source).viaMat(flow)(Keep.right).to(subSink.sink).run()(subFusingMaterializer)
+        } match {
+          case Success(matVal) =>
+            innerMatValue.success(matVal)
+            Initializing.upstreamFailure match {
+              case OptionVal.Some(ex) =>
+                subSource.fail(ex)
+              case OptionVal.None =>
+                if (isClosed(in))
+                  subSource.complete()
+            }
+            Initializing.downstreamCause match {
+              case OptionVal.Some(cause) =>
+                subSink.cancel(cause)
+              case OptionVal.None =>
+                //todo: should this be invoked before and independently of checking downstreamCause?
+                // in most case if downstream pulls and then closes, the pull is 'lost'. is it possible for some flows to actually care about this? (non-eager broadcast?)
```

Comment:
It's fine like this. It might change behavior in upstream stages but that's expected. Cancellation can happen at any time.